### PR TITLE
[CI] Add Buildkite

### PR DESCRIPTION
**Pull request by simon-mo**
This PR adds basic setup for GPU CI environment. It should enable us to run our tests on L4 GPUs. As developer, you can add new tests to `.buildkite/test-pipeline.yaml`. 

Currently, I have all tests enabled, in addition to benchmarks. However, I don't want to block the merge of this PR for debugging model output (test_models.py) and tuning memory (test_attention.py). I marked those tests as "soft fail" for now. 

Please reference the latest build here: https://buildkite.com/vllm/ci/builds/182 as example. The end to end time of the build on worst case (fresh docker build, slow machine starts) is about 1hr, on the best case (docker cached, machine available) is about 15 minutes. Of course, if there are too many PRs submitted at the same time, it might needs to wait and queue up a bit. We are capped at 10 GPU machines due to budget reason. The full infrastructure setup is described and maintained in a separate repo https://github.com/vllm-project/buildkite-ci. 

Code change in the PR is mostly in `.buildkite` directory and associated `Dockerfile` and `setup.py`. Everything else is done in order to make existing test pass. 

Future work includes:
* Fix test models
* Fix test kernels
* Migrate lint, docs, and wheels to CPU only machines
* Add tests for chat models (test chat template coverage)
* Benchmark real models (instead of just OPT) on A100

### CODE REVIEW

**Comment by simon-mo** [comment]
note to self: The one last blocker is the memory requirements for huggingface model, we can run models on 2xL4 in vLLM but huggingface doesn't have a good TP strategy that's simple to use. currently i'm trying accelerate to do offload, but if it doesn't work we might go A100

**Comment by simon-mo** [comment]
todos: migrate lint, docs, and wheels to CPU only machines

**Comment by simon-mo** [comment]
I soft failed the kernels and models tests. The models ran successfully with bfloat16 but the some output doesn't match :(. The kernel is too difficult to tune.

File: .buildkite/test-pipeline.yaml [hunk 1]
```diff
@@ -0,0 +1,41 @@
+# In this fail, you can add more tests to run either by adding a new step or
```
Comment (simon-mo) [inline .buildkite/test-pipeline.yaml:1]:
```suggestion
# In this file, you can add more tests to run either by adding a new step or
```

File: tests/async_engine/test_api_server.py [hunk 2]
```diff
@@ -50,12 +55,15 @@ def test_api_server(api_server):
...
+        max_tries = 10
         while not result:
             try:
                 for r in pool.map(_query_server, prompts):
                     result = r
                     break
-            except requests.exceptions.ConnectionError:
+            except requests.exceptions.ConnectionError as err:
+                if max_tries == 0:
+                    raise RuntimeError("Server did not start") from err
```
Comment (zhuohan123) [inline tests/async_engine/test_api_server.py:66]:
```suggestion
                if max_tries == 0:
                    raise RuntimeError("Server did not start") from err
                max_tries -= 1
```

File: tests/distributed/test_comm_ops.py [hunk 3]
```diff
@@ -2,10 +2,9 @@
 
 Run `pytest tests/distributed/test_comm_ops.py --forked`.
 """
-from multiprocessing import Process, set_start_method
-
 import pytest
 import torch
+import ray
```
Comment (zhuohan123) [inline tests/distributed/test_comm_ops.py:7]: Just curious, why change to ray?

File: tests/models/test_models.py [hunk 4]
```diff
@@ -21,7 +21,8 @@
 
 
 @pytest.mark.parametrize("model", MODELS)
-@pytest.mark.parametrize("dtype", ["float"])
+# half is required to get this working on CI's L4 GPU
+@pytest.mark.parametrize("dtype", ["half"])
```
Comment (zhuohan123) [inline tests/models/test_models.py:25]: Let's keep this as float, since otherwise the test will fail on a100s.

Comment (zhuohan123) [reply tests/models/test_models.py:25]: We can implement a simpler test for L4 in another PR.

**Review by zhuohan123 — APPROVED** [review]
Thanks for the hard work Simon! Left some small comments.

**Comment by zhuohan123** [comment]
@simon-mo Let’s expedite this PR so that we can have CI working for all other PRs?

File: tests/distributed/test_comm_ops.py (hunk 3 above)
Comment (simon-mo) [reply tests/distributed/test_comm_ops.py:7]: in the code below, i added a comment saying ray gives way better log for debugging purpose (i couldn't figure the failures from multiprocessing)

Comment (zhuohan123) [reply tests/distributed/test_comm_ops.py:7]: gotcha! makes a lot of sense.

**Pull request merged**